src/pint/bayesian.py

```diff
@@ -0,0 +1,206 @@
+from copy import deepcopy
+
+import numpy as np
+from scipy.linalg import cho_factor, cho_solve
+from scipy.stats import norm, uniform
+
+from pint.models.priors import Prior, UniformUnboundedRV
+from pint.residuals import Residuals
+
+
+class BayesianTiming:
+    """A wrapper around the PINT API that provides lnprior, prior_transform,
+    lnlikelihood, and lnposterior functions. This interface can be used to
+    draw posterior samples using the sampler of your choice.
+
+    Parameters
+    ----------
+    model : :class:`pint.models.TimingModel`
+        The best-fit values stored in this object are not used.
+    toas : a :class:`pint.toa.TOAs` instance. Contains the input toas.
+    use_pulse_numbers : bool, optional
+        How to handle phase wrapping. If True, will use the pulse numbers
+        from the toas object while creating :class:`pint.residuals.Residuals`
+        objects. Otherwise will use the nearest integer.
+    prior_info : dict, optional
+        A dict containing the prior information on free parameters. This parameter
+        supersedes any priors present in the model.
+
+    Notes
+    -----
+    > The `prior` attribute of each free parameter in the `model` object should
+      be set to an instance of :class:`pint.models.priors.Prior`.
+
+    > The parameters of BayesianTiming.model will change for every likelihood function
+      call. These parameters in general will not be the best-fit values. Hence, it is NOT
+      a good idea to save it as a par file.
+
+    > Currently, only uniform and normal distributions are supported in prior_info. More
+      general priors should be set directly in the TimingModel object before creating the
+      BayesianTiming object.
+      Here is an example prior_info object:
+        prior_info = {
+            "F0" : {
+                "distr" : "normal",
+                "mu"    : 1,
+                "sigma" : 0.00001
+            },
+            "EFAC1" : {
+                "distr" : "uniform",
+                "pmin"  : 0.5,
+                "pmax"  : 2.0
+            }
+        }
+    """
+
+    def __init__(self, model, toas, use_pulse_numbers=False, prior_info=None):
+        # Make a deep copy to not mess up the original model.
+        self.model = deepcopy(model)
+        self.toas = toas
+
+        self.param_labels = self.model.free_params
+        self.params = [getattr(model, par) for par in self.param_labels]
+        self.nparams = len(self.param_labels)

```

Comment:
Again, might we want only free parameters?  Or is there a way to use a delta function prior?